### PR TITLE
HIVE-24750. Create a single copy task for external tables within default DB location.

### PR DESCRIPTION
**Pull request by ayushtkn**

### CODE REVIEW

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosExternalTables.java [hunk 1]
```diff
@@ -1375,6 +1376,205 @@ public void differentCatalogIncrementalReplication() throws Throwable {
...
+    DistributedFileSystem fs = primary.miniDFSCluster.getFileSystem();
+    fs.mkdirs(externalTableLocation, new FsPermission("777"));
+
+    Path externalTablePartitionLocation =
+        new Path("/" + testName.getMethodName() + "/" + primaryDbName + "/" + "part/");
+    fs.mkdirs(externalTableLocation, new FsPermission("777"));
+
+    List<String> withClause = Arrays.asList(
+        "'distcp.options.update'='','hive.repl.external.warehouse.single.copy"
+            + ".task'='true'",
```
Comment (pkumarsinha) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosExternalTables.java:1401]: nit: accommodate the config in single line itself. Will improve readability

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosExternalTables.java [hunk 2]
```diff
@@ -1375,6 +1376,205 @@ public void differentCatalogIncrementalReplication() throws Throwable {
...
+  public void testDatabaseLevelCopyLazy() throws Throwable {
+    testDatabaseLevelCopy(true);
+  }
+
+  @Test
+  public void testDatabaseLevelCopyAtSource() throws Throwable {
+    testDatabaseLevelCopy(false);
+  }
+
+  public void testDatabaseLevelCopy(boolean isAtTarget) throws Throwable {
```
Comment (pkumarsinha) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosExternalTables.java:1389]: nit: isAtTarget -> runCopyTasksOnTarget ?

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosExternalTables.java [hunk 3]
```diff
@@ -1375,6 +1376,205 @@ public void differentCatalogIncrementalReplication() throws Throwable {
...
+            .run("insert into a values(1,2)")
+            .run("create external table b (id int)")
+            .run("insert into b values(5)")
+            .run("create external table c (place string) partitioned by (country "
+                + "string)")
+            .run("insert into table c partition(country='india') values "
+                + "('bangalore')")
+            .run("ALTER TABLE c ADD PARTITION (country='france') LOCATION '"
+                + externalTablePartitionLocation.toString() + "'")
+            .run("insert into c partition(country='france') values('paris')")
```
Comment (pkumarsinha) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosExternalTables.java:1421]: After this line, can you please add assertions on table and partition locations

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosExternalTables.java [hunk 4]
```diff
@@ -1375,6 +1376,205 @@ public void differentCatalogIncrementalReplication() throws Throwable {
...
+        .verifyResult("5")
+        .run("select place from c where country='india'")
+        .verifyResult("bangalore")
+        .run("select place from c where country='france'")
+        .verifyResult("paris");
+
+    // Check the task copied post bootstrap, It should have the database loc,
+    // the table 'a' since that is outside of the default location, and the
+    // 'c', since its partition is out of the default location.
+    assertExternalFileInfo(Arrays.asList(primaryDbName.toLowerCase(), "a", "c"),
```
Comment (pkumarsinha) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosExternalTables.java:1439]: What is the source path considered and is printed as a part of C here? Is it table location for C or the partition location for partition, i.e externalTablePartitionLocation

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosExternalTables.java [hunk 5]
```diff
@@ -1375,6 +1376,205 @@ public void differentCatalogIncrementalReplication() throws Throwable {
...
+            .run("insert into b values(6)")
+            .run("insert into table c partition(country='india') values "
+                + "('delhi')")
+            .run("insert into c partition(country='france') values('lyon')")
+            .run("create external table newin (id int)")
+            .run("insert into newin values(1)")
+            .run("create external table newout(id int) row format delimited "
+                + "fields terminated by ',' location '" + externalTableLocation
+                .toUri() + "'")
+            .run("insert into newout values(2)")
```
Comment (pkumarsinha) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosExternalTables.java:1460]: After this line, can you please add assertions on table and partition locations

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplExternalTables.java [hunk 6]
```diff
@@ -149,10 +150,15 @@ void dataLocationDump(Table table, FileList fileList, HiveConf conf)
             "only External tables can be writen via this writer, provided table is " + table
                 .getTableType());
       }
-      Path fullyQualifiedDataLocation =
-          PathBuilder.fullyQualifiedHDFSUri(table.getDataLocation(), FileSystem.get(hiveConf));
-      write(lineFor(table.getTableName(), fullyQualifiedDataLocation, hiveConf));
-      dirLocationToCopy(fileList, fullyQualifiedDataLocation, conf);
+      Path fullyQualifiedDataLocation = PathBuilder
```
Comment (pkumarsinha) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplExternalTables.java:153]: nit: Can accommodate in a single line

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplExternalTables.java [hunk 7]
```diff
@@ -180,6 +195,14 @@ void dataLocationDump(Table table, FileList fileList, HiveConf conf)
       }
     }
 
+    void dbLocationDump(String dbName, Path dbLocation, FileList fileList,
+        HiveConf conf) throws Exception {
+      Path fullyQualifiedDataLocation = PathBuilder
+          .fullyQualifiedHDFSUri(dbLocation, FileSystem.get(hiveConf));
+      write(lineFor(dbName, fullyQualifiedDataLocation, hiveConf));
```
Comment (pkumarsinha) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplExternalTables.java:202]: lineFor assumes that the entry is for table name . We are passing here dbName. What are the side effect?

Comment (ayushtkn) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplExternalTables.java:200]: Nopes, the name isn't used anywhere, it writes to `_external_tables_info` which is anyway going to get deleted

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosExternalTables.java (hunk 4 above)
Comment (ayushtkn) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosExternalTables.java:1467]: That would be the partition location, the name used is the table name, That logic was already there

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosExternalTables.java [hunk 8]
```diff
@@ -1375,6 +1376,245 @@ public void differentCatalogIncrementalReplication() throws Throwable {
...
+        new Path("/" + testName.getMethodName() + "/" + primaryDbName + "/" + "a/");
+    DistributedFileSystem fs = primary.miniDFSCluster.getFileSystem();
+    fs.mkdirs(externalTableLocation, new FsPermission("777"));
+
+    Path externalTablePartitionLocation =
+        new Path("/" + testName.getMethodName() + "/" + primaryDbName + "/" + "part/");
+    fs.mkdirs(externalTableLocation, new FsPermission("777"));
+
+    List<String> withClause = Arrays.asList(
+        "'distcp.options.update'='','hive.repl.external.warehouse.single.copy.task'='true'",
```
Comment (pkumarsinha) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosExternalTables.java:1401]: Sorry, missed it last time. Please use const. for the config. hive.repl.external.warehouse.single.copy.task

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosExternalTables.java [hunk 9]
```diff
@@ -1375,6 +1376,245 @@ public void differentCatalogIncrementalReplication() throws Throwable {
...
+            .run("ALTER TABLE c ADD PARTITION (country='france') LOCATION '"
+                + externalTablePartitionLocation.toString() + "'")
+            .run("insert into c partition(country='france') values('paris')")
+            .dump(primaryDbName, withClause);
+
+    Database primaryDb = primary.getDatabase(primaryDbName);
+
+    // Confirm the a table is outside the db location.
+    Table aTable = primary.getTable(primaryDbName, "a");
+    new Path(aTable.getSd().getLocation());
```
Comment (pkumarsinha) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosExternalTables.java:1428]: Is this line added by mistake? You can remove this as it doesn't seem to be doing anything.

**Review by pkumarsinha — COMMENTED** [review]
Just two minor comments. Changes looks fine otherwise.

**Review by pkumarsinha — APPROVED** [review]

**Pull request merged**